### DISCUSSION TITLE
Buffer management questions

### DISCUSSION THREAD

**Discussion by kupiakos** [tpm-rs/tpm-rs#184] [Design & Requirments]
General questions about buffer management that we should answer sooner rather than later:

- Can the request and response buffers alias?
- Does TPM manage its own request/response buffer (e.g. as a `static` or on the stack), or is it provided externally?
- Does the response have to be written to a single contiguous buffer, or can it write through something like a `Writer` trait?
- Do we support directly marshalling/unmarshalling MMIO memory, i.e. memory that requires volatile access to correctly use?

MMIO memory is currently unergonomic to work with: you cannot create a reference to MMIO memory. See [this](https://github.com/rust-lang/unsafe-code-guidelines/issues/265) and [this](https://github.com/rust-lang/unsafe-code-guidelines/issues/411) discussion under unsafe-code-guidelines. So, the "perfect" copying `TpmRead` trait that is capable of working with volatile and nonvolatile and also avoids unnecessary zeroing looks something like:

```rust
// Implemented for &[u8], &mut [u8], VolatilePtr<[u8], impl Readable>
trait TpmRead {
    // Out<'_, [u8]> is a write-only buffer: a `&mut [u8]` or `&mut [MaybeUninit<u8>]`
    fn read_into(&mut self, output: Out<'_, [u8]>) -> Result<&mut [u8], Error>;
}

// Implemented for &[u8], &mut [u8], VolatilePtr<[u8], impl Writeable>
trait TpmWrite {
    fn write_from(&mut self, input: &[u8]) -> Result<(), Error>;
}

// Writes to an `Out<T>` instead of returning `Result<T>` because `T` could be large and unpacking the `Result` could be expensive even with inlining.
impl<T: AsBytes + FromBytes> Marshal for T {
    fn try_unmarshal<R: TpmRead>(reader: &mut R, storage: Out<'_, T>) -> Result<&mut Self,  Err> {
        Ok(unsafe { &mut *(reader.read_into(out_type_to_out_bytes(storage))?).cast() }
    }

    fn try_marshal<W: TpmWrite>(&self, writer: &mut W) -> Result<(),  Err> {
        writer.write_from(self.as_bytes())
    }
}
```

**Comment by vsukhoml** [comment]
I'd say TPM shall receive a buffer allocated by higher-level entity,  use only stack and its state and update same buffer.  Not sure how MMIO is related here? Is it a CRB method vs. FIFO?
